### PR TITLE
Blueprints: add blueprints slice for client global state 

### PR DESCRIPTION
**Pull request by amirfefer**
Made on top of  #1654, first two commits unrelated. 

This PR adds blueprints slice for managing blueprints client state globally. Multiple components use these data as props and args, and for keeping it consistent when a change occurs,  redux slice gives a clean solution.

### CODE REVIEW

**Comment by lucasgarfield** [comment]
This had not occurred to me but I think it will probably end up being a good idea, I like it.

Haven't done an in-depth review yet but two things immediately:
1. Can you please move `BlueprintsSlice.tsx` into `/store`, use `pascalCase` for the title, and use a `.ts.` extension? Just to be consistent with, for instance, `wizardSlice.ts`.
2. Would it be possible to just rebase this onto `main`? Then I think we could get it in pretty quickly. Might take a while to get first two commits it is rebased onto in this PR reviewed and merged.

File: src/store/BlueprintSlice.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+type blueprintsState = {
+  selectedBlueprint: string | undefined;
+  searchInput?: string;
+  offset?: number;
+  limit?: number;
+};
+
+const initialState: blueprintsState = {
+  selectedBlueprint: undefined,
+  searchInput: undefined,
```
Comment (lucasgarfield) [inline src/store/BlueprintSlice.ts:14]: Let's initialize this as `''`.

File: src/store/BlueprintSlice.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import { PayloadAction, createSlice } from '@reduxjs/toolkit';
+
+import { RootState } from '.';
+
+type blueprintsState = {
+  selectedBlueprint: string | undefined;
+  searchInput?: string;
+  offset?: number;
```
Comment (lucasgarfield) [inline src/store/BlueprintSlice.ts:8]: Would we ever want the offset and limit to be undefined? I suppose if we wanted the query to use whatever the default values are... but given that we'll be implementing pagination I think we will need fine control. Can we make these 
```
offset: number;
limit: number;
```
instead?

File: src/store/BlueprintSlice.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+};
+
+const initialState: blueprintsState = {
+  selectedBlueprint: undefined,
+  searchInput: undefined,
+  offset: 0,
+  limit: 10,
+};
+
+export const selectSelectedBlueprint = (state: RootState) =>
```
Comment (lucasgarfield) [inline src/store/BlueprintSlice.ts:19]: Please change to `selectSelectedBlueprintId`.

File: src/Components/ImagesTable/ImagesTable.tsx [hunk 4]
```diff
@@ -59,15 +64,14 @@ import {
...
-type ImageTableProps = {
-  selectedBlueprint?: string | undefined;
-};
-
-const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
+const ImagesTable = () => {
   const [page, setPage] = useState(1);
   const [perPage, setPerPage] = useState(10);
+  const selectedBlueprint = useAppSelector(selectSelectedBlueprint);
+  const blueprintsSearchQuery = useAppSelector(selectBlueprintSearchInput);
```
Comment (lucasgarfield) [inline src/Components/ImagesTable/ImagesTable.tsx:71]: Would you mind changing this to `blueprintSearchInput`? I find it easier to understand the code when the name of the variable in the React component is the same as it is in the store.

File: src/Components/Blueprints/DeleteBlueprintModal.tsx [hunk 5]
```diff
@@ -7,25 +7,27 @@ import {
...
   DeleteBlueprintModalProps
-> = ({
-  onDelete,
-  selectedBlueprint,
-  isOpen,
-  onClose,
-}: DeleteBlueprintModalProps) => {
+> = ({ onDelete, isOpen, onClose }: DeleteBlueprintModalProps) => {
+  const selectedBlueprint = useAppSelector(selectSelectedBlueprint);
+  const blueprintsSearchQuery = useAppSelector(selectBlueprintSearchInput);
```
Comment (lucasgarfield) [inline src/Components/Blueprints/DeleteBlueprintModal.tsx:27]: Would you mind changing this to `blueprintSearchInput`? I find it easier to understand the code when the name of the variable in the React component is the same as it is in the store.

File: src/Components/Blueprints/BlueprintsSideBar.tsx [hunk 6]
```diff
@@ -141,30 +124,37 @@ const BlueprintsSidebar = ({
...
   );
 };
 
-const BlueprintSearch = ({
-  filter,
-  setFilter,
-  blueprintsTotal,
-}: blueprintSearchProps) => {
+const BlueprintSearch = ({ blueprintsTotal }: blueprintSearchProps) => {
+  const blueprintsSearchQuery = useAppSelector((state) =>
```
Comment (lucasgarfield) [inline src/Components/Blueprints/BlueprintsSideBar.tsx:136]: Would you mind changing this to `blueprintSearchInput`? I find it easier to understand the code when the name of the variable in the React component is the same as it is in the store.

File: src/Components/Blueprints/BlueprintsSideBar.tsx [hunk 7]
```diff
@@ -49,27 +47,13 @@ type emptyBlueprintStateProps = {
...
-    [setBlueprintsSearchQuery]
-  );
-  React.useEffect(() => {
-    return () => {
-      debouncedSearch.cancel();
-    };
-  }, [debouncedSearch]);
+const BlueprintsSidebar = () => {
+  const selectedBlueprint = useAppSelector(selectSelectedBlueprint);
+  const blueprintsSearchQuery = useAppSelector(selectBlueprintSearchInput);
```
Comment (lucasgarfield) [inline src/Components/Blueprints/BlueprintsSideBar.tsx:52]: Would you mind changing this to blueprintSearchInput? I find it easier to understand the code when the name of the variable in the React component is the same as it is in the store.

**Comment by lucasgarfield** [comment]
Wow, this is really nice! The components are going to be so much easier to work with now. I know the variable name comments are a bit nitpicky, but I think that the comments about the types in the store are actually something really important to get correct from the beginning.

**Comment by ezr-ondrej** [comment]
@amirfefer can you rebase this pls? I guess this would be nice to get in prioritized, to base other work on this.

File: src/store/BlueprintSlice.ts (hunk 1 above)
Comment (amirfefer) [reply src/store/BlueprintSlice.ts:14]: When the backend get empty string (`''`)  in the search  param,  it will do an empty searching...

Comment (ezr-ondrej) [reply src/store/BlueprintSlice.ts:14]: Yeah, lets not xD or fix the backend first :)

File: src/store/BlueprintSlice.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
+import { PayloadAction, createSlice } from '@reduxjs/toolkit';
+
+import { RootState } from '.';
+
+type blueprintsState = {
+  selectedBlueprint: string | undefined;
+  searchInput?: string;
```
[the remark this answers is not in the document]
Comment (amirfefer) [reply src/store/BlueprintSlice.ts:7]: There's a reason why undefined is needed here for empty search, as you can see in the endpoint implementation  - https://github.com/osbuild/image-builder/blob/main/internal/v1/handler_blueprints.go#L192

File: src/store/BlueprintSlice.ts (hunk 2 above)
Comment (amirfefer) [reply src/store/BlueprintSlice.ts:8]: The default values for offset and limit in the endpoint are `0` and `100` respectively, we can have a default values here as well, but we need to keep it consistent.

File: src/store/BlueprintSlice.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import { PayloadAction, createSlice } from '@reduxjs/toolkit';
+
+import { RootState } from '.';
+
+type blueprintsState = {
+  selectedBlueprint: string | undefined;
```
[the remark this answers is not in the document]
Comment (amirfefer) [reply src/store/BlueprintSlice.ts:6]: I  kept the previous name, to have less changes in code, tho `selectedBlueprintId` is better

File: src/store/BlueprintSlice.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (lucasgarfield) [reply src/store/BlueprintSlice.ts:7]: Okay, as long as there's a good reason!

File: src/store/BlueprintSlice.ts (hunk 1 above)
Comment (lucasgarfield) [reply src/store/BlueprintSlice.ts:14]: Fair enough :smile:

File: src/store/BlueprintSlice.ts (hunk 2 above)
Comment (lucasgarfield) [reply src/store/BlueprintSlice.ts:8]: Okay that's fine.

**Comment by ezr-ondrej** [comment]
/retest

npm acting out

**Review by lucasgarfield — DISMISSED** [review]
Thanks Amir! I think this is going to be the last "big" PR, we're so close!

**Review by lucasgarfield — APPROVED** [review]

**Pull request merged**